feat(server): Add initial minidump scrubbing

flub opened the pull request:
Implements initial data scrubbing of UTF-8 strings in heap and stack memory streams in minidumps. All sections are marked as `pii = "maybe"`, which makes it effectively disabled for current PII rules.

Commit:
Extremely rough and inefficient version of zeroing out memory

This also zeroes out the stack, which is entirely undesirable.

untitaker committed:
initial work on attachment scrubbing

untitaker committed:
ref(pii): Split out regex mapping into separate module

Move regex mapping into separate module. We need to share this logic with attachment scrubbing once it's there.

Also change usage of `Option<BTreeSet<u8>>` into dedicated enum.

untitaker committed:
ref(pii): Split out regex mapping into separate module

Move regex mapping into separate module. We need to share this logic with attachment scrubbing once it's there.

Also change usage of `Option<BTreeSet<u8>>` into dedicated enum.

untitaker committed:
initial work

Commit:
Only zero-out non-stack memory regions

This also removes some of the manual work to read the minidump as the
existing API does actually expose enough to get to all the
information.

untitaker committed:
wip

untitaker committed:
wip

untitaker committed:
remove leftover binary type

untitaker committed:
fix compilation

untitaker committed:
finish implementation and add more value types

Commit:
Allow access to both stack and non-stack memory

This also puts some indirection in place and documents things a bit
more.

untitaker committed:
wip testing

untitaker committed:
remove valuetype inheritance and fix up tests

Commit:
First step in editing code names

This is pretty basic, but works.

untitaker committed:
wire up minidump example with pii processor and fix some bugs

untitaker committed:
add testcase for bytes regexes

untitaker committed:
fix formatting

Commit:
Expose the linux environment

Also refactor things a bit, use MINIDUMP_LOCATION_DESCRIPTOR instead
of our own custom types.  Add a trait to make it's slicing easy.

Commit:
Use this new trait in mem_desc as well

flub commented on examples/edit_minidump.rs:
```diff
@@ -0,0 +1,209 @@
+            &filename,
+            data.desc_mut_slice(&raw_desc),
+            AttachmentBytesType::PlainAttachment,
+        );
```
@untitaker this is extremely untested, it just compiles.  The format of these is just a bunch of bytes, which are "unix string" (aka any corrupted encoding possible) separated by NULL bytes.

untitaker replied:
I wonder whether we can get the locale settings out of the minidump to figure out how to decode paths, env, and whatever this is.

Possibly we can find locale settings in the env, so we could try to lossy-decode the env as ascii first, parse out locale, then parse everything (except the memory) as that encoding.

flub replied:
Yes, locale can normally be parsed out of the environ and used to parse out the encoding (my past self seems to think https://hg.sr.ht/~flub/syskit/browse/syskit/_utils.py?rev=tip#L109 is the way to get this, there's probably rust libs though).

Doing this gets back to how to reverse an encoding to modify the minidump though.

Commit:
Start integrating minidump scrubbing in relay

Commit:
Add LinuxCmd and LinuxEnviron scrubbing

This needs access also to the raw data to do pointer arithmetic.  So
this refactors this a bit to have one struct which contains both the
Minidump struct and the raw data.

Commit:
Move things and rebase

Commit:
Fix up Cargo.toml and actually add minidump module

jan-auer committed:
Merge branch 'master' into feat/minidump-edit-example

* master:
  fix(danger): Comment only on missing changelogs (#723)
  test: Require any python3 and support python3.8 (#722)

jan-auer committed:
ref: Clean up implementation

jan-auer committed:
ref: Move utility into tools folder

jan-auer committed:
Merge branch 'master' into feat/minidump-edit-example

* master:
  feat(server): Support chunked formdata for Crashpad (#721)

jan-auer committed:
meta: Changelog

jan-auer committed:
build: Fix docker build

untitaker reviewed (changes requested):
This looks fine but we need to figure out how to deploy this without enabling. Right now this will match on selectors `*` and `**`, and the former cannot be solved with selector specificness at the moment. Also if somebody were to have configured a selector named `'UE4Minidump.dmp'`, this would (and should) apply as far as I understand.

I think the sanest way forward would be to have a featureflag inside of the projectconfig that is attached to a featureflag in Sentry.

jan-auer committed:
ref: Mark minidumps and attachments as pii=maybe

jan-auer committed:
ref: Better error messages

jan-auer committed:
fix: Clippy suggests static over const

jan-auer committed:
ref: Format
